<a href="https://colab.research.google.com/github/ckdlsqpf/projectmanager/blob/main/%ED%86%B5%EA%B3%843.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [14]:
import pandas as pd
df = pd.read_excel('car.xlsx')
df.head()

mileage model  price  year  my_car_damage  other_car_damage
0    63608    K3    970  2017              0            564596
1    69336    K3   1130  2015        1839700           1140150
2    36000    K3   1380  2016         446520           2244910
3    19029    K3   1390  2017         889000           4196110
4    97090    K3    760  2015        2339137           2029570

In [12]:
# 회귀분석
from statsmodels.formula.api import ols

m = ols('price~year', data=df).fit()
m.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  price   R-squared:                       0.687
Model:                            OLS   Adj. R-squared:                  0.686
Method:                 Least Squares   F-statistic:                     597.3
Date:                Thu, 14 Sep 2023   Prob (F-statistic):           1.39e-70
Time:                        00:42:48   Log-Likelihood:                -1820.2
No. Observations:                 274   AIC:                             3644.
Df Residuals:                     272   BIC:                             3652.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept  -2.135e+05   8771.400    -24.341      0.000   -2.31e+05   -1.96e+05
year         106.4745      4.357     24.439      0.000      97.897     115.052
==============================================================================
Omnibus:                        9.748   Durbin-Watson:                   1.525
Prob(Omnibus):                  0.008   Jarque-Bera (JB):                9.774
Skew:                           0.447   Prob(JB):                      0.00754
Kurtosis:                       3.236   Cond. No.                     1.57e+06
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.57e+06. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

- year가 1년 증가하면 106만원씩 증가한다.
- 귀무가설: 모집단에서 기울기는 0이다.(연식이 높든 낮든 가격과 무관하다.)
 - p-val < 0.05 귀무가설 기각.

In [15]:
new_df = pd.DataFrame({'year': [2014,2015,2016,2017]})
new_df

year
0  2014
1  2015
2  2016
3  2017

In [17]:
m.predict(new_df)

0     930.990601
1    1037.465097
2    1143.939593
3    1250.414089
dtype: float64

In [18]:
import pingouin as pg
pg.corr(df.price, df.year)

n         r         CI95%         p-val       BF10  power
pearson  274  0.828908  [0.79, 0.86]  1.388002e-70  1.004e+67    1.0

In [19]:
0.828908 ** 2

0.687088472464

- 68%의 예측력이 나오겠다!

In [20]:
m = ols('price~mileage + year', data=df).fit()
m.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  price   R-squared:                       0.749
Model:                            OLS   Adj. R-squared:                  0.747
Method:                 Least Squares   F-statistic:                     403.5
Date:                Thu, 14 Sep 2023   Prob (F-statistic):           5.62e-82
Time:                        01:11:34   Log-Likelihood:                -1790.2
No. Observations:                 274   AIC:                             3586.
Df Residuals:                     271   BIC:                             3597.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept  -1.688e+05   9597.865    -17.592      0.000   -1.88e+05    -1.5e+05
mileage       -0.0023      0.000     -8.143      0.000      -0.003      -0.002
year          84.3822      4.761     17.724      0.000      75.009      93.755
==============================================================================
Omnibus:                       11.272   Durbin-Watson:                   1.598
Prob(Omnibus):                  0.004   Jarque-Bera (JB):               11.786
Skew:                           0.435   Prob(JB):                      0.00276
Kurtosis:                       3.523   Cond. No.                     8.41e+07
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 8.41e+07. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [30]:
mileage = 50000
year = 2015
price = -168800 + (-0.0023 * mileage) + (84.3822 * year)
price

1115.1330000000016

In [31]:
m = ols('price ~ scale(mileage) + scale(year)', data=df).fit()
m.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  price   R-squared:                       0.749
Model:                            OLS   Adj. R-squared:                  0.747
Method:                 Least Squares   F-statistic:                     403.5
Date:                Thu, 14 Sep 2023   Prob (F-statistic):           5.62e-82
Time:                        01:58:59   Log-Likelihood:                -1790.2
No. Observations:                 274   AIC:                             3586.
Df Residuals:                     271   BIC:                             3597.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==================================================================================
                     coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------
Intercept        853.6606     10.112     84.419      0.000     833.752     873.569
scale(mileage)  -100.2044     12.306     -8.143      0.000    -124.431     -75.978
scale(year)      218.1006     12.306     17.724      0.000     193.874     242.327
==============================================================================
Omnibus:                       11.272   Durbin-Watson:                   1.598
Prob(Omnibus):                  0.004   Jarque-Bera (JB):               11.786
Skew:                           0.435   Prob(JB):                      0.00276
Kurtosis:                       3.523   Cond. No.                         1.91
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

**price = 853.6606 -100 scale(mileage) + 218 scale(year)**
* 표준화했을 때 절편이 853만원이라는 뜻:
평균적인 mileage와 평균적인 year의 차가
853만원 정도 한다
* mileage가 1표준편차(43015Km) 증가하면
가격은 100만원 떨어진다
* year가 1표준편차(2.6년) 증가하면 가격은 218만원 상승
* year가 가격에는 영향이 더 크다고 할 수 있음

In [32]:
m = ols('price ~ model', data=df).fit()
m.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  price   R-squared:                       0.011
Model:                            OLS   Adj. R-squared:                  0.007
Method:                 Least Squares   F-statistic:                     3.039
Date:                Thu, 14 Sep 2023   Prob (F-statistic):             0.0824
Time:                        02:17:07   Log-Likelihood:                -1977.9
No. Observations:                 274   AIC:                             3960.
Df Residuals:                     272   BIC:                             3967.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
===============================================================================
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
Intercept     833.4146     23.144     36.009      0.000     787.850     878.980
model[T.K3]    80.3970     46.121      1.743      0.082     -10.402     171.196
==============================================================================
Omnibus:                       13.893   Durbin-Watson:                   0.528
Prob(Omnibus):                  0.001   Jarque-Bera (JB):               15.007
Skew:                           0.573   Prob(JB):                     0.000551
Kurtosis:                       3.002   Cond. No.                         2.48
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

- Intercept	833.4146 (Avante 예상가격)
- model[T.K3]	80.3970 (K3와 Avante의 예상가격 차이)

**price = 833 + 80 * model[T.K3]**

- Avante: `model[T.K3] = 0` 833만원으로 예상
- K3: `model[T.K3] = 1` 913만원으로 예상

- p-value > 0.05이므로 귀무가설: "둘의 차이 = 0"을 기각할 수 없음
- 또는 신뢰구간 -10~+171(0도 포함됨)
- 따라서 모집단에서도 K3가 Avante보다 비싸다고 하기에는 증거가 부족하다

In [36]:
avante = df.price[df.model == 'Avante']
k3 = df.price[df.model =='K3']
pg.ttest(k3, avante, correction=False)
# 독립표본 t 검정에서 보정을 안하면 회귀분석과 결과가 같음
# 무엇을 보정하나? 이분산성(heteroscedasticity): 가격대가 퍼져있는 정도가 다르다

T  dof alternative     p-val           CI95%   cohen-d   BF10  \
T-test  1.743179  272   two-sided  0.082432  [-10.4, 171.2]  0.242614  0.627   

         power  
T-test  0.4119

- 두 집단의 차이가 나느냐, 안 나느냐는 t 검정이 더 정확(왜? 보정을 하니까~)
- 그러면 회귀분석을 왜 하냐? 예측을 하기 위해서

In [34]:
m = ols('price ~ year + mileage + model', data=df).fit()
m.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  price   R-squared:                       0.751
Model:                            OLS   Adj. R-squared:                  0.749
Method:                 Least Squares   F-statistic:                     271.9
Date:                Thu, 14 Sep 2023   Prob (F-statistic):           2.98e-81
Time:                        02:32:21   Log-Likelihood:                -1788.8
No. Observations:                 274   AIC:                             3586.
Df Residuals:                     270   BIC:                             3600.
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
===============================================================================
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
Intercept   -1.713e+05   9670.404    -17.714      0.000    -1.9e+05   -1.52e+05
model[T.K3]   -40.4089     23.626     -1.710      0.088     -86.925       6.107
year           85.6030      4.798     17.843      0.000      76.158      95.048
mileage        -0.0023      0.000     -8.186      0.000      -0.003      -0.002
==============================================================================
Omnibus:                        9.726   Durbin-Watson:                   1.623
Prob(Omnibus):                  0.008   Jarque-Bera (JB):                9.901
Skew:                           0.405   Prob(JB):                      0.00708
Kurtosis:                       3.458   Cond. No.                     8.50e+07
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 8.5e+07. This might indicate that there are
strong multicollinearity or other numerical problems.
"""